### ISSUE TITLE
Add usage / key usage for sidetree methods

### ISSUE THREAD

**Issue by OR13** [w3c/did-core#158]
https://github.com/decentralized-identity/sidetree/issues/362#issuecomment-578825320

We need to formalize this concept of KeyUsage in a way that plays well with proof purposes.

For sidetree, there are some keys which are used to recover. They should be represented as follows I believe;

```
 {
      "publicKeyHex": "03f4c82cee043eeb8dbed2bf02a6b9b066c742557816a738f74a6e4dfd2bbc6469",
      "type": "EcdsaSecp256k1VerificationKey2019",
      "id": "#recovery",
      "keyUsage": "sign",
      "controller": "did:elem:EiChaglAoJaBq7bGWp6bA5PAQKaOTzVHVXIlJqyQbljfmg"
    },

...

 "sidetreeRecoveryKeys": ["#recovery"],
```

This would unify a couple concepts while avoiding setting a pattern of arbitrarily annotating public key objects (in a way that breaks the did core spec).

**Comment by tplooker** [comment]
@OR13 how come recovery doesn't just become another verification method?

To relate to your example above I think the DID doc would look something like the following

```
{
  "@context": "https://www.w3.org/ns/did/v1",
  "id": "did:example:123456789abcdefghi",
  "publicKey": [{
    "id": "did:example:123456789abcdefghi#keys-1",
    "type": "EcdsaSecp256k1VerificationKey2019",
    "controller": "did:example:123456789abcdefghi",
    "publicKeyHex": "03f4c82cee043eeb8dbed2bf02a6b9b066c742557816a738f74a6e4dfd2bbc6469",
  }],
  "recovery": ["did:example:123456789abcdefghi#keys-1"],
  "assertionMethod": ["did:example:123456789abcdefghi#keys-1"],
  "authentication": ["did:example:123456789abcdefghi#keys-1"]
}
```

Depending on what you meant by `"keyUsage": "sign"`, e.g is that digital signing for the purposes of authenticating or digital signing for the purposes of asserting something? In either case the assertionMethod or authentication element could be included in the did document to express that?

**Comment by OR13** [comment]
My intention is to get @dlongley 's note in the linked comment some better visibility, I will repeat it here in full: 

> I'm not sure how usage is being defined here, but if it is like Web Crypto API "KeyUsage", then the above statement is not true.

"KeyUsage" from that spec refers to the cryptographic operations that a particular key can perform, which is not the same thing as a user authorizing a verification method to be used to verify a proof that was made for a particular purpose.

For example:

Key A, with "KeyUsage" "sign", can be used to sign proofs for the purpose of authentication. A proof of this sort could be sent to a service to "log you in"/"establish a session", i.e., generally used to prove to another party that they are talking with whomever is in control of a particular key.

Key B, also with "KeyUsage" "sign", can be used to sign proofs for the purpose of assertionMethod. A proof of this sort is merely a method of making an assertion, like a set of claims about some subject, aka a Verifiable Credential, and presenting such a proof should NOT be used to "log you in"/"establish a session". It would be a problem if Alice were to give Bob a VC that says Bob is her friend and Bob could turn around and use that to log into a service as Alice -- because it is her proof on the VC.

Both keys use the "sign" cryptographic method here, but for very different purposes. Also note that both keys can mathematically (or according only to the cryptographic protocol) be used to sign proofs for either purpose. The math will work. This can be dangerous if Alice is running software that isn't sufficiently protecting her from accidentally signing the hash of a message that could be used for authentication -- when she thought it was only for making an assertion she intended to share with another party.

BUT, a DID Document allows Alice to specify which verification methods are permitted to verify proofs based on the purpose of those proofs. Alice can use her DID Document to say that the verification method associated with key A (i.e., its public key) can only be used to verify proofs for the purpose of authentication and the verification for key B can only be used to verify proofs for the purpose of assertionMethod. That way verification libraries can be written just once that will check the proofPurpose and verificationMethod Alice puts on her proofs against the authorization relations she has put in her DID Document.

If a proof says "proofPurpose": "authentication", "verificationMethod": ["did:foo:123#key"] then the verification library will only follow in the authentication relation (aka property/attribute/JSON key) in her DID Document to look for "did:foo:123#key" when verifying. If it's not there, it's not authorized to be used to verify a proof for that purpose. The same applies if the proof were to say "proofPurpose": "assertionMethod" instead; the verification library would only look under the assertionMethod relation for the verificationMethod "did:foo:123#key". This ensures that this authorization is coming from Alice (it is a relation between her and the verification method). So, note that this is not the same thing as what cryptographic operations the key can technically perform (which is a relation between the key and its cryptographic functions).


IMO, the DID Core spec needs to start by explaining how we intend to handle proof purposes and key usages in a way that does not create confusion with existing concepts... like what happened when sidetree just randomly added "usage" to a public key, without defining it in the did core spec.

**Comment by OR13** [comment]
@selfissued @csuwildcat  this issue is related to sidetree did methods, and the extent to which our recovery concept is "standard" or "method specific". 

Currently, it relies on a method specific term 'usage" which is not supported by the core context, however, we've fixed this issue in element today....

I really don't like that sidetree just added "usage" without any consideration for existing interoperability here... while we have fixed it in element as I mentioned, I'd love to see this cleaned up in did core, and have what we have done in sidetree not just be a 1 off.

**Comment by OR13** [comment]
@csuwildcat IMO, you should be responsible for sorting this out :) 

I'd prefer we not be doing what we are with sidetree and the "usage" property, which currently causes breaking issues.

**Comment by OR13** [comment]
This is also blocked by the did core registry... is usage to be defined there?

**Comment by ChristopherA** [comment]
I would also like to see a way to officially have a key usage as ALL (maybe “*”) but also put a note that this is discouraged. In the case of BTCR the establishing transaction has all possible key usages and authority, and only in subsequent updates of the extended DID Document is it attenuated into different keys for different key usage. 

Currently I have to keep updating my template for creating an implicit DID document as we define key usages, where I put the same key into every use.

This may also be a good pattern for other methods with an establishing transaction, such as did:peer or did:github.

— Christopher Allen

**Comment by OR13** [comment]
Microsoft and the Sidetree WG, have removed this as a requirement, and it is no longer needed.
